### ISSUE TITLE
Design Matrix Class Hierarchy

### ISSUE THREAD

**Issue by bkochuna** [bkochuna/ners570f20-Lab06#8]
To be filled out in class

**Comment by mhig-umich** [comment]
![image](https://user-images.githubusercontent.com/53484799/95659387-add76200-0aee-11eb-90b8-3ee3d1df63af.png)

**Comment by mhig-umich** [comment]
![image](https://user-images.githubusercontent.com/53484799/95659450-23dbc900-0aef-11eb-8c14-d045f4418531.png)

I created a matrix hierarchy similar to the one made in class but I further separated by one additional subclass which is based on if the elements are stored in either a 1D or 2D format. I'm not sure if all my notation is correct but this is how I imagine it in my head!

**Comment by bkochuna** [comment]
@mhig-umich did a really nice design. We discussed this in lecture on 10/12. What we developed in lecture was:
![Class Hierarchy](https://user-images.githubusercontent.com/5183217/96207278-d3be9580-0f38-11eb-94b4-b5ff5694fffe.png)

A few notes for correctness in @mhig-umich's design.
- The generalization arrows are backwards
- The attributes and methods defined in the parent class should not be included in the child classes

Also for clarity: the figure in this comment is also not _completely correct_. Here we assume that the `SparseMatrix_CSC` follows the same pattern as the others.
